pkg/controller/traffic/traffic_controller.go

```diff
@@ -311,97 +311,160 @@ func (c *Controller) processTrafficTargetOnCluster(
 
 		return err
 	}
 
-	informerFactory, err := c.clusterClientStore.GetInformerFactory(spec.Name)
+	appName := tt.Labels[shipper.AppLabel]
+	releaseName := tt.Labels[shipper.ReleaseLabel]
+
+	appPods, endpoints, err := c.getClusterObjects(spec.Name, tt.Namespace, appName)
 	if err != nil {
-		cond := trafficutil.NewClusterTrafficCondition(
+		operationalCond = trafficutil.NewClusterTrafficCondition(
 			shipper.ClusterConditionTypeOperational,
 			corev1.ConditionFalse,
-			ServerError,
+			InternalError,
 			err.Error(),
 		)
-		diff.Append(trafficutil.SetClusterTrafficCondition(status, *cond))
 
 		return err
 	}
 
-	cond := trafficutil.NewClusterTrafficCondition(
+	operationalCond = trafficutil.NewClusterTrafficCondition(
 		shipper.ClusterConditionTypeOperational,
 		corev1.ConditionTrue,
 		"",
 		"",
 	)
-	diff.Append(trafficutil.SetClusterTrafficCondition(status, *cond))
 
-	achievedWeight, err := shifter.SyncCluster(spec.Name, clientset, informerFactory)
-	if err != nil {
-		var reason string
-
-		switch err.(type) {
-		case shippererrors.UnexpectedObjectCountFromSelectorError:
-			reason = MissingService
-		case shippererrors.KubeclientError:
-			reason = ServerError
-		case shippererrors.TargetClusterMathError:
-			reason = InternalError
-		default:
-			reason = UnknownError
-		}
-
-		cond := trafficutil.NewClusterTrafficCondition(
-			shipper.ClusterConditionTypeReady,
-			corev1.ConditionFalse,
-			reason,
-			err.Error(),
-		)
-		diff.Append(trafficutil.SetClusterTrafficCondition(status, *cond))
-
-		return err
-	}
+	trafficStatus := buildTrafficShiftingStatus(
+		spec.Name, appName, releaseName,
+		clusterReleaseWeights,
+		endpoints, appPods)
 
-	status.AchievedTraffic = achievedWeight
+	// achievedTraffic is used by the defer at the top of this func
+	achievedTraffic = trafficStatus.achievedTrafficWeight
 
-	var readyCond *shipper.ClusterTrafficCondition
-	if achievedWeight == spec.Weight {
+	if trafficStatus.ready {
 		readyCond = trafficutil.NewClusterTrafficCondition(
 			shipper.ClusterConditionTypeReady,
 			corev1.ConditionTrue,
 			"",
 			"",
 		)
+
+		return nil
+	}
+
+	if trafficStatus.podsToShift != nil {
+		err := shiftPodLabels(clientset, trafficStatus.podsToShift)
+		if err != nil {
+			readyCond = trafficutil.NewClusterTrafficCondition(
+				shipper.ClusterConditionTypeReady,
+				corev1.ConditionFalse,
+				InternalError,
+				err.Error(),
+			)
+
+			return err
+		} else {
+			readyCond = trafficutil.NewClusterTrafficCondition(
+				shipper.ClusterConditionTypeReady,
+				corev1.ConditionFalse,
+				InProgress,
+				"",
+			)
+		}
 	} else {
 		readyCond = trafficutil.NewClusterTrafficCondition(
 			shipper.ClusterConditionTypeReady,
 			corev1.ConditionFalse,
 			PodsNotReady,
-			fmt.Sprintf("weight expected: %d, weight achieved: %d", spec.Weight, achievedWeight),
+			fmt.Sprintf(
+				"%d out of %d pods are not ready. this might require intervention, check the CapacityTarget for more information",
```

Comment:
We might consider using this guide we propose in release controller: "try using kubectl get ct ..." for the sake of consistency here.